mmeds/util.py

```diff
@@ -1125,27 +1131,166 @@ def parse_barcodes(forward_barcodes, reverse_barcodes, forward_mapcodes, reverse
+    map_df[('#SampleID', '#q2:types')] = map_df[('#SampleID', '#q2:types')].str.replace('@', '')
+
+    # make sure this column is first
+    map_df.set_index(('#SampleID', '#q2:types'), inplace=True)
+    map_df.reset_index(inplace=True)
+
+    # add this column to the end of the mapping file, so it passes Qiime1 mapping file validation
+    map_df[('Description', 'categorical')] = np.nan
+
+    map_df.to_csv(f'{output_dir}/{file_name}_qiime_barcode_mapfile.tsv', index=None, header=True, sep='\t')
+    if ret_dicts:
+        ret_val = (map_df, results_dict, full_dict)
+    else:
+        ret_val = map_df
+
+    return ret_val
+
+
+def validate_demultiplex(demux_file, for_barcodes, rev_barcodes, map_file, log_dir, is_gzip, get_read_counts=False,
+                         on_chimera=True):
+    """
+    Calls qiime1 script to validate a gzipped, demultiplex fastq file.
+    source_dir: where the data is located.
+    demux_file: path to gzipped, demultiplexed, fastq file.
+    for_barcodes: path to gzipped, forward barcode file
+    rev_barcodes: path to reverse barcode file
+    map_file: path to a qiime mapping file
+    log_dir: path to log dir, where the results will be saved
+    """
+    # TODO: generalize for single barcodes
+
+    # Allows us to test that the correct barcodes are in the resulting demultiplexed file.
+    barcode_return = create_barcode_mapfile(Path(demux_file).parent, for_barcodes, rev_barcodes,
+                                            Path(demux_file).stem, map_file, get_read_counts)
+    if get_read_counts:
+        map_df = barcode_return[0]
+        matched_barcodes = barcode_return[1]
+        all_barcodes = barcode_return[2]
+    else:
+        map_df = barcode_return
+
+    new_env = setup_environment('qiime/1.9.1')
+
+    test_file = f'{Path(demux_file).parent}/{Path(demux_file).stem}_test.fastq'
+    map_file = f'{Path(demux_file).parent}/{Path(demux_file).stem}_qiime_barcode_mapfile.tsv'
+
+    gunzip_demux_file = ['gunzip', f'{demux_file}.gz']
+    gzip_demux_file = ['gzip', demux_file]
+    create_fastq_copy = ['cp', f'{demux_file}', f'{test_file}']
+
+    # to create the fasta file, on every line mod4 = 1, replace @ with >
+    # then on every line mod4 = 2, print the line. then skip all other lines.
+    create_fasta_file = ['sed', '-n', '-i', '1~4s/^@/>/p;2~4p', f'{test_file}']
+
+
+    # Call qiime1 validate demultiplex script
+    if on_chimera:
+        validate_demux_file = f'ml python/2.7.9-UCS4; validate_demultiplexed_fasta.py -b -a\
+                            -i {test_file}\
+                            -o {log_dir}\
+                            -m {map_file};\
+                            rm {test_file}'
+    else:
+        validate_demux_file = f'validate_demultiplexed_fasta.py -b -a\
+                            -i test_file\
+                            -o {log_dir}\
+                            -m {map_file};\
+                            rm {test_file}'
```

Comment:
```suggestion
    if on_chimera:
        validate_demux_file = [
                             'ml python/2.7.9-UCS4;',
                             'validate_demultiplexed_fasta.py -b -a',
                             '-i {}'.format(test_file),
                             '-o {}'.format(log_dir),
                             '-m {};'.format(map_file),
                             'rm {};'.format(test_file)
                 ]
```
do something like this instead, it's more consistent with the codebase and those backslashes look clunky. then when you need to run it do you also need `validate_demux_file.join(' ')`? I would assume you do but it doesn't look like you needed that with the other commands you wrote as lists

also use their full option names instead of the single letter if possible

Reply:
booo no we like fstrings over .format. .format is the old way! backslashes in this case are needed to split across multiple lines.
I wrote the commands as lists because that's how I found David did it, sometimes that caused issues though and then I did it as a single string rather than a list of strings.
I'll find the full option names and update

Reply:
lol, it's fine to use fstrings then, but still use a list like this instead of one big fstring

Reply:
nah need one big fstring :)
having it as a list does break the code, I spent toooo long trying to make sure this worked